### PR TITLE
[CIS-234] Implement guest users

### PR DESCRIPTION
**Pull request by mantergo**
(on behalf of @evsaev )

**This PR:** 

- Provides implementation for `ChatClient.setGuestUser`;
- Introduces `guestUserToken` endpoint;
- Introduces `GuestUserTokenRequestPayload` for `guestUserToken` request;
- Introduces `GuestUserTokenPayload` for `guestUserToken` response;
- Adds generic dummy method to `CurrentUserPayload`.

### CODE REVIEW

File: StreamChatClient_v3/APIClient/Endpoints/Requests/GuestUser.swift [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+/// Type that serves guest-user params
+struct GuestUser<ExtraData: UserExtraData>: Encodable {
+    private enum CodingKeys: String, CodingKey {
+        case id
+    }
+
+    /// User's identifier
+    let id: UserId
+    /// User's extra data
+    let extraData: ExtraData?
```
Comment (VojtaStavik) [inline StreamChatClient_v3/APIClient/Endpoints/Requests/GuestUser.swift:16]: `ExtraData` is not an optional field in v3

File: StreamChatClient_v3/ChatClient.swift [hunk 2]
```diff
@@ -222,8 +222,34 @@ public class Client<ExtraData: ExtraDataTypes> {
...
+
+            self.apiClient.request(endpoint: .token(userId: userId, extraData: extraData)) {
+                switch $0 {
+                case let .success(payload):
+                    self.currentToken = payload.token
+                    self.connect(completion: completion)
+                case let .failure(error):
+                    completion?(error)
+                }
+            }
```
Comment (VojtaStavik) [inline StreamChatClient_v3/ChatClient.swift:251]: What a nice retain cycle we have here :)

File: StreamChatClient_v3/ChatClient_Tests.swift [hunk 3]
```diff
@@ -368,6 +368,76 @@ class ChatClient_Tests: StressTestCase {
...
+
+            // New user id is set
+            Assert.willBeEqual(client.currentUserId, newUser.id)
+
+            // Make sure `guest` endpoint is called
+            Assert.willBeEqual(self.testEnv.apiClient!.request_endpoint,
+                               AnyEndpoint(.token(userId: newUser.id, extraData: newUserExtraData)))
+
+            // Make sure the completion is not called yet
+            Assert.willBeFalse(setUserCompletionCalled)
```
Comment (VojtaStavik) [inline StreamChatClient_v3/ChatClient_Tests.swift:404]: This should probably be `Assert.staysFalse` or you can move it after the `AsertAsync` block and do just `XCTAssertFalse(setUserCompletionCalled)`. The `AssertAsync` blocks the execution until all conditions are met.

File: StreamChatClient_v3/ChatClient.swift (hunk 2 above)
Comment (evsaev) [reply StreamChatClient_v3/ChatClient.swift:251]: Actually, there is no retain cycle here. String ref to `ChatClient` captured by a closure will be alive until the network request is finished as the lifetime of this closure is limited to the lifetime of the data-task. So it doesn't lead to a memory leak.

At the same time, you are Absolutely right that it doesn't make sense to keep `ChatClient` alive if the calling code doesn't have any/ drop all strong references to this object 👍

Comment (VojtaStavik) [reply StreamChatClient_v3/ChatClient.swift:251]: Sorry, bad wording from my side. It's not a retain cycle. What I meant is the situation you described. It's not a big deal really, but it seems to me it's safer when the completion is not called, once the client is released.

File: StreamChatClient_v3/APIClient/Endpoints/Requests/GuestUser.swift [hunk 4]
```diff
@@ -0,0 +1,24 @@
+//
+// Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+/// Type that serves guest-user params
+struct GuestUser<ExtraData: UserExtraData>: Encodable {
```
[the remark this answers is not in the document]
Comment (evsaev) [reply StreamChatClient_v3/APIClient/Endpoints/Requests/GuestUser.swift]: fixed in 0f7845d545017e8c34d738d2123f49509eb45efe

File: StreamChatClient_v3/APIClient/Endpoints/Requests/GuestUser.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (evsaev) [reply StreamChatClient_v3/APIClient/Endpoints/Requests/GuestUser.swift]: fixed in 0f7845d545017e8c34d738d2123f49509eb45efe

File: StreamChatClient_v3/APIClient/Endpoints/Payloads/GuestUserTokenPayload.swift [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+/// Type that describes response from **guest** endpoint
+struct GuestUserTokenPayload<ExtraData: UserExtraData>: Decodable {
+    private enum CodingKeys: String, CodingKey {
+        case user
+        case token = "access_token"
+    }
+
+    /// The user's payload
+    let user: UserPayload<ExtraData>
```
[the remark this answers is not in the document]
Comment (evsaev) [reply StreamChatClient_v3/APIClient/Endpoints/Payloads/GuestUserTokenPayload.swift]: fixed in 17cb6fd52167cba4549b5939af4f3277f3fedf46

File: StreamChatClient_v3/APIClient/Endpoints/GuestEndpoints.swift [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+import Foundation
+
+extension Endpoint {
+    /// Builds the endpoint to obtain an access-token for a guest user.
+    /// - Parameters:
+    ///   - userId: The user's identifier
+    ///   - extraData: The user's extra data
+    /// - Returns: The endpoint that expects **GuestUserTokenPayload<ExtraData>** payload in a response
+    static func token<ExtraData: UserExtraData>(userId: UserId, extraData: ExtraData?) -> Endpoint<GuestUserTokenPayload<ExtraData>> {
```
[the remark this answers is not in the document]
Comment (evsaev) [reply StreamChatClient_v3/APIClient/Endpoints/GuestEndpoints.swift]: fixed in a9c667de705d886456e90a6cb99f81cbb779680e

File: StreamChatClient_v3/APIClient/Endpoints/Requests/GuestUser.swift (hunk 1 above)
Comment (evsaev) [reply StreamChatClient_v3/APIClient/Endpoints/Requests/GuestUser.swift]: fixed in 0f7845d545017e8c34d738d2123f49509eb45efe

File: StreamChatClient_v3/ChatClient.swift (hunk 2 above)
Comment (evsaev) [reply StreamChatClient_v3/ChatClient.swift]: fixed in c2f9b54ca1e776672e3feafe43f9cca993534e08

File: StreamChatClient_v3/ChatClient_Tests.swift (hunk 3 above)
Comment (evsaev) [reply StreamChatClient_v3/ChatClient_Tests.swift]: fixed in c2f9b54ca1e776672e3feafe43f9cca993534e08

**Comment by VojtaStavik** [comment]
@evsaev Any blockers here? I think we can merge.

**Comment by VojtaStavik** [comment]
ping again @evsaev

**Pull request merged**